### PR TITLE
Improvement: add metrics to prom client

### PR DESCRIPTION
**Pull request by OGKevin**
This pr add `pkg/extprom/http/instrument_client.go` with code to instrument an HTTP client.
The approach is to add metrics clients to the HTTP config.
Not sure if this is the right approach or if another clever design can be used, this was the most obvious and backward compatible way I could come up with.

fixes: https://github.com/thanos-io/thanos/issues/4545

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Verification

```
# HELP thanos_rule_query_http_client_in_flight_requests A gauge of in-flight requests.
# TYPE thanos_rule_query_http_client_in_flight_requests gauge
thanos_rule_query_http_client_in_flight_requests 0
# HELP thanos_rule_query_http_client_request_duration_seconds A histogram of request latencies.
# TYPE thanos_rule_query_http_client_request_duration_seconds histogram
thanos_rule_query_http_client_request_duration_seconds_bucket{le="0.005"} 0
thanos_rule_query_http_client_request_duration_seconds_bucket{le="0.01"} 1
thanos_rule_query_http_client_request_duration_seconds_bucket{le="0.025"} 2
thanos_rule_query_http_client_request_duration_seconds_bucket{le="0.05"} 2
thanos_rule_query_http_client_request_duration_seconds_bucket{le="0.1"} 2
thanos_rule_query_http_client_request_duration_seconds_bucket{le="0.25"} 2
thanos_rule_query_http_client_request_duration_seconds_bucket{le="0.5"} 2
thanos_rule_query_http_client_request_duration_seconds_bucket{le="1"} 2
thanos_rule_query_http_client_request_duration_seconds_bucket{le="2.5"} 2
thanos_rule_query_http_client_request_duration_seconds_bucket{le="5"} 2
thanos_rule_query_http_client_request_duration_seconds_bucket{le="10"} 2
thanos_rule_query_http_client_request_duration_seconds_bucket{le="+Inf"} 2
thanos_rule_query_http_client_request_duration_seconds_sum 0.018484452
thanos_rule_query_http_client_request_duration_seconds_count 2
# HELP thanos_rule_query_http_client_request_total 
# TYPE thanos_rule_query_http_client_request_total counter
thanos_rule_query_http_client_request_total{code="200",method="post"} 2
```

### CODE REVIEW

File: pkg/promclient/promclient.go [hunk 1]
```diff
@@ -68,17 +71,36 @@ type Client struct {
...
+}
+
+type clientMetrics struct {
+	requestCount *prometheus.SummaryVec
+}
+
+func newClientMetrics(reg prometheus.Registerer) *clientMetrics {
+	var m clientMetrics
+
+	m.requestCount = promauto.With(reg).NewSummaryVec(prometheus.SummaryOpts{
```
Comment (bwplotka) [inline pkg/promclient/promclient.go:85]: We already have HTTP client metrics middleware for it. Let me double check if we don't instrument this right now...

Comment (bwplotka) [reply pkg/promclient/promclient.go:85]: Actually we don't and we only have server HTTP middleware (and both client and server for GRPC), so this what we miss and we need to add :https://github.com/thanos-io/thanos/blob/8f0afeea960bd9c162edaa29642892716c396976/pkg/extprom/http/instrument_server.go#L35

File: pkg/extprom/http/instrument_client.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	dnsLatencyHistogram      *prometheus.HistogramVec
+	tlsLatencyHistogram      *prometheus.HistogramVec
+	requestDurationHistogram *prometheus.HistogramVec
+}
+
+// NewClientMetrics creates a new instance of ClientMetrics.
+// It will also register the metrics with the included register.
+// This ClientMetrics should be re-used for diff clients with the same purpose.
+// e.g. 1 ClientMetrics should be used for all the clients that talk to Alertmanager.
+func NewClientMetrics(reg prometheus.Registerer, namespace string) *ClientMetrics {
```
Comment (bwplotka) [inline pkg/extprom/http/instrument_client.go:28]: I wonder if we can stick to per code metric naming. 

Essentially we can have `http_client_*` metrics  (no thanos prefix or anything) and use some `extprom.WrapRegistererWithLabel(reg, "client" "alertmanager") as reg parameter here.

Having stable name for all instances is powerful - allows greater metric stability and overall better dashboards/alerts/rules patterns reusability!


```suggestion
func NewClientMetrics(reg prometheus.Registerer) *ClientMetrics {
```

**Review by bwplotka — APPROVED** [review]
Nice! I love it. LGTM, just one minor nit.

File: pkg/extprom/http/instrument_client.go (hunk 2 above)
Comment (OGKevin) [reply pkg/extprom/http/instrument_client.go:28]: aah this is a good suggestion 👌🏾

File: pkg/extprom/http/instrument_client.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+			Namespace: namespace,
+			Subsystem: "http_client",
+			Name:      "tls_duration_seconds",
+			Help:      "Trace tls latency histogram.",
+			Buckets:   []float64{0.025, .05, .1, .5, 1, 5, 10},
+		},
+		[]string{"event"},
+	)
+
+	m.requestDurationHistogram = promauto.With(reg).NewHistogramVec(
```
[the remark this answers is not in the document]
Comment (OGKevin) [reply pkg/extprom/http/instrument_client.go:67]: This actually needs to be Vec due to prom client expecting it.

https://github.com/prometheus/client_golang/blob/efe7aa73021f40ba96d167c144941fa5fb7f43e2/prometheus/promhttp/instrument_client.go#L88

**Comment by OGKevin** [comment]
The metrics now look like this:

```
# HELP http_client_in_flight_requests A gauge of in-flight requests.
# TYPE http_client_in_flight_requests gauge
http_client_in_flight_requests{client="alertmanager"} 0
http_client_in_flight_requests{client="query"} 0
# HELP http_client_request_duration_seconds A histogram of request latencies.
# TYPE http_client_request_duration_seconds histogram
http_client_request_duration_seconds_bucket{client="query",le="0.025"} 24
http_client_request_duration_seconds_bucket{client="query",le="0.05"} 24
http_client_request_duration_seconds_bucket{client="query",le="0.1"} 24
http_client_request_duration_seconds_bucket{client="query",le="0.5"} 24
http_client_request_duration_seconds_bucket{client="query",le="1"} 24
http_client_request_duration_seconds_bucket{client="query",le="5"} 24
http_client_request_duration_seconds_bucket{client="query",le="10"} 24
http_client_request_duration_seconds_bucket{client="query",le="+Inf"} 24
http_client_request_duration_seconds_sum{client="query"} 0.105682592
http_client_request_duration_seconds_count{client="query"} 24
# HELP http_client_request_total 
# TYPE http_client_request_total counter
http_client_request_total{client="query",code="200",method="post"} 24
```

File: pkg/extprom/http/instrument_client.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/extprom/http/instrument_client.go:63]: Do we want to add labels to this histogram? For example, `code` and `method`? https://github.com/prometheus/client_golang/blob/efe7aa73021f40ba96d167c144941fa5fb7f43e2/prometheus/promhttp/instrument_server.go#L293:6

File: pkg/extprom/http/instrument_client.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	m.inFlightGauge = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Subsystem: "http_client",
+		Name:      "in_flight_requests",
+		Help:      "A gauge of in-flight requests.",
+	})
+
+	m.requestTotalCount = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Subsystem: "http_client",
+		Name:      "request_total",
+		Help:      "",
```
Comment (bwplotka) [inline pkg/extprom/http/instrument_client.go:40]: We need some help (:

File: pkg/extprom/http/instrument_client.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply pkg/extprom/http/instrument_client.go:63]: Yea, I don't think it expects if we have `prometheus.Histogram` type not `prometheus.HistogramVec` (:

File: pkg/http/http.go [hunk 5]
```diff
@@ -35,6 +37,9 @@ type ClientConfig struct {
 	ProxyURL string `yaml:"proxy_url"`
 	// TLSConfig to use to connect to the targets.
 	TLSConfig TLSConfig `yaml:"tls_config"`
+	// ClientMetrics contains metrics that will be used to instrument
+	// the client that will be created with this config.
+	ClientMetrics *extpromhttp.ClientMetrics
```
Comment (bwplotka) [inline pkg/http/http.go:42]: we could do `yaml:"-"`` to make sure parser will not parse it (:

**Review by bwplotka — DISMISSED** [review]
Sweet, all good except two nits. Good job!

**Comment by bwplotka** [comment]
e2e fail looks unrelated, but let's double check. Retrying.

File: pkg/extprom/http/instrument_client.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (OGKevin) [reply pkg/extprom/http/instrument_client.go:63]: Added the labels.

File: pkg/extprom/http/instrument_client.go (hunk 4 above)
Comment (OGKevin) [reply pkg/extprom/http/instrument_client.go]: added

**Review by bwplotka — APPROVED** [review]

**Comment by bwplotka** [comment]
Retried, will monitor if this time e2e fails again or not.

**Pull request merged**